gtime/metrics/metrics.py

```diff
@@ -112,5 +112,224 @@ def max_error(
+
+
+def r_square(
+    y_true: Union[pd.DataFrame, List, np.ndarray],
+    y_pred: Union[pd.DataFrame, List, np.ndarray],
+) -> float:
+    """Compute the R squared (Coefficient of Determination) between two vectors.
+
+    Parameters
+    ----------
+    y_true : array-like, shape (length, 1), required
+        The first vector.
+
+    y_pred : array-like, shape (length, 1), required
+        The second vector.
+
+    Returns
+    -------
+    r_square : float
+        The R squared between the two vectors.
+
+    Examples
+    --------
+    >>> from gtime.metrics import r_square
+    >>> y_true = [0, 1, 2, 3, 4, 5]
+    >>> y_pred = [1.1, 2.3, 0.4, 3.9, 3.1, 4.6]
+    >>> r_square(y_true, y_pred)
+    0.586
+
+    """
+    y_true, y_pred = _convert_to_ndarray(y_true, y_pred)
+    _check_input(y_true, y_pred)
+    
+    ss_res = sum((y_true - y_pred) ** 2)
+    ss_tot = sum((y_true - np.mean(y_true)) ** 2)
+    if not np.any(ss_tot):
+        if not np.any(ss_res):
+            return 1.0
+        else:
+            return 0.0 
+    if np.isnan(ss_res / ss_tot):
+        return np.NINF
+    r_square = 1 - (ss_res / ss_tot)
+    return r_square
+
+
+def mae(
+    y_true: Union[pd.DataFrame, List, np.ndarray],
+    y_pred: Union[pd.DataFrame, List, np.ndarray],
+) -> float:
+    """Compute the Mean Absolute Error(also called, Mean Absolute Deviation(MAD)) between two vectors.
+
+    Parameters
+    ----------
+    y_true : array-like, shape (length, 1), required
+        The first vector.
+
+    y_pred : array-like, shape (length, 1), required
+        The second vector.
+
+    Returns
+    -------
+    mae_value : float
+        The mean absolute error between the two vectors.
+
+    Examples
+    --------
+    >>> from gtime.metrics import mae
+    >>> y_true = [0, 1, 2, 3, 4, 5]
+    >>> y_pred = [1.1, 2.3, 0.4, 3.9, 3.1, 4.6]
+    >>> mae(y_true, y_pred)
+    1.033
+
+    """
+    y_true, y_pred = _convert_to_ndarray(y_true, y_pred)
+    _check_input(y_true, y_pred)
+    
+    mae_value = np.mean(np.abs(y_pred - y_true))
+    return mae_value
+
+
+def mape(
+    y_true: Union[pd.DataFrame, List, np.ndarray],
+    y_pred: Union[pd.DataFrame, List, np.ndarray],
+) -> float:
+    """Compute the Mean Absolute Percentage Error(MAPE) between two vectors.
+
+    Parameters
+    ----------
+    y_true : array-like, shape (length, 1), required.
+        The first vector.
+
+    y_pred : array-like, shape (length, 1), required.
+        The second vector.
+
+    Returns
+    -------
+    mape_value : float
+        The mean absolute percentage error between the two vectors.
+
+    Examples
+    --------
+    >>> from gtime.metrics import mape
+    >>> y_true = [1, 1, 2, 3, 4, 5]
+    >>> y_pred = [1, 2.3, 0.4, 3.9, 3.1, 4.6]
+    >>> mape(y_true, y_pred)
+    45.08
+
+    """
+    y_true, y_pred = _convert_to_ndarray(y_true, y_pred)
+    _check_input(y_true, y_pred)
+    
+    ratio_list = np.abs((y_pred - y_true)/y_true)
+    if (y_true == 0).any():
```

Comment:
I think all of this can be rewritten as:

    mape_value = np.mean(np.abs((y_pred - y_true)/y_true))
    if np.isna(mape_value):
        raise ValueError("MAPE can not be calculated due to Zero/Zero")
    return mape_value * 100

Correct me if I'm wrong.

Reply:
True, I have made the code change and tested it. It works fine and faster than the previous implementation.